### DISCUSSION TITLE
Concept of "Order"/"Priority" in Optimization?

### DISCUSSION THREAD

**Discussion by kmcentush** [cvxpy/cvxpy#1824] [Q&A]
Hi all,

I've used CVXPY in different contexts over the years for American fantasy football drafts. Specifically, I would try to optimize a pick for a given team based on the remaining players to choose from as well as constraints like needing at least one kicker on your team. I was most successful in doing this when I allowed the optimizer to pick my entire time at once (as if I had all my picks in a row), then in reality I would draft the top player it suggested. Eventually when it was my turn to draft again, I would do the same scheme - optimize as if I had all my picks in a row, pick one player, etc.

I opted for this approach as opposed to only optimizing for a single "next best pick" because of the overall roster constraints. For example, I need to have at least one kicker. Kickers score way fewer points than quarterbacks. Without optimizing an entire team at once, the routine would almost certainly suggest as many quarterbacks as possible until the instance I add the kicker constraint when it would be forced to take a kicker.

I wanted to try to go further than this - rather than optimizing one team at a team, optimize _all_ teams at once. However, since there is an order to the draft, I was wondering if there was a way to encode this "priority" into the optimization. So ultimately, the optimizer would spit out the remaining picks for every team, keeping in mind that my optimal team will look slightly different after the person before me makes an optimal draft choice. I believe this approach would do a better job mimicking the reality of my final team as opposed to what I've done in the past (again, acting like I have all my picks in a row). If the differences in results between these two approaches is not clear, please let me know. I can come up with an example.

Questions - 
1. So, is there a way to do something like this? Effectively, if there are N picks remaining in the draft, is there a way to have the optimization problem solve for all N picks, but make pick 2 after optimizing pick 1, 3 after 2, etc? But with the overall constraints still applied all at once? In essence, encode the order/priority of the decision but still be part of a single optimization routine.

2. If not, do you have any suggestions for better applying constraints (that apply to the final requirements) but may not be relevant early on? ex: I need one kicker on my team, but I don't want to wait until the final round to add that constraint to the solver. But as soon as I add it the solver will pick a kicker...

Thank you for any and all ideas!

**Comment by kmcentush** [comment]
Perhaps the solution would involve Leximin ordering - https://github.com/cvxpy/cvxpy/discussions/1505 and , where I give the earlier draft picks higher "values" for their decision somehow along the lines of this example - https://github.com/erelsgl/cvxpy_leximin/blob/main/examples/leximin_allocation.py.

**Comment by SteveDiamond** [comment]
First off, this is a great application, thanks for telling us about it! Would you have any interest in writing an example with a simplified version? 

I guess though in terms of your question, is it possible to optimize all the teams at once? Isn't it a multi-player game?

**Reply by SteveDiamond** [reply]
If it's a multi-player game then generally that's not going to yield a convex problem.

**Reply by kmcentush** [reply]
While it may be multiplayer in principle, if we make an assumption about every player's behavior I think it may actually be feasible. I just posted an example below - please let me know your thoughts.

**Comment by kmcentush** [comment]
@SteveDiamond

Sure thing! So here's a small example. In the past, I've set my objective to be to maximize the minimum the points that my team scores per week. In the actual game, every week your team "plays" an opponent team. Whichever team scores the highest points wins.

Anyways, with the "maximize the minimum points per week" objective in mind, here is a dummy example. Say there are three teams (human players) total, and each team must have exactly one quarterback and one running back - so 2 picks each, 6 total in the draft. An example breakdown of the expected points per week is as follows.
![image](https://user-images.githubusercontent.com/20652336/178867806-658b88d7-96ad-42d4-bcff-7abfac1600eb.png)

If the draft order is as follows (snake-style), what is the optimal team for everyone?
1. Team 1
2. Team 2
3. Team 3
4. Team 3
5. Team 2
6. Team 1

**Method 1: Optimize as if you get all the picks in a row.**
I've used this method in the past, so I'm confident that it is convex.
![image](https://user-images.githubusercontent.com/20652336/178867899-bdf13176-328b-4ae1-8dde-142976d18827.png)

The `Optimal Team` column is what the optimizer outputs each pick. Note that it changes as the draft goes on as players are taken by other teams.

Using this method, where it assumes you can always pick the best players sequentially, Team 1 ends up with a minimum points per week of 29, which is below both Team 2 and 3's 30. 

**Method 2: Optimize with knowledge of opponent's picks** (assuming they would act optimally, i.e. to also maximize their minimum points per week):
Code that can do this is what I'm specifically asking about in this discussion thread.
![image](https://user-images.githubusercontent.com/20652336/178868155-a17e9a67-1b4c-4ef5-9b56-c653177c6256.png)

Again, the `Optimal Team` column is what the optimizer would output each pick. Note that it is static throughout the whole draft - that is because we assume all players are acting optimally. In a real draft with non-optimal behavior by some teams, the optimal team may change pick-to-pick - but theoretically non-optimal behavior by some teams should only increase the floor of our objective, i.e. helping make us better.

Using this method, Team 1 recognizes that it would actually be sub-optimal (again, relative to our objective to maximize the minimum points per week) to pick the supposed best player in the draft - QB1. Instead, they take QB2, knowing that later they're going to end up with RB3. After the draft concludes, Team 1 has actually increased its minimum points per week from 29 in Method 1 to 30 here - tying it with Team 3. Team 2 now has the highest minimum at 31, but ultimately Team 1's objective was to maximize it's own minimum points. Any changes in other team's minimums are consequences of this optimal behavior by Team 1.

Hopefully that helps clarify things. Method 2 is somewhat related to game theory but in a very simplistic way - assume all actors behave the same way, and optimize for that.

**Comment by kmcentush** [comment]
For anyone curious in helping out, here's the code implementation for the setup and method 1 solution based on the dummy example above. This is in line with what I've used in previous fantasy football seasons.

I think that the method 2 code (whatever that would look like, should it be possible) might happen in a single optimization problem as opposed to the looping over multiple problems like in method 1.

```
#!/usr/bin/env python
# coding: utf-8

# In[ ]:


import cvxpy as cp  # note: also need cvxopt installed
import numpy as np
import pandas as pd


# # Setup

# In[ ]:


# Define players and their point projections
players = pd.DataFrame(
    [
        ["QB1", "QB", 20, 15],
        ["QB2", "QB", 18, 16],
        ["QB3", "QB", 17, 16],
        ["RB1", "RB", 16, 14],
        ["RB2", "RB", 12, 16],
        ["RB3", "RB", 13, 14],
    ],
    columns=["uid", "position", "week1", "week2"],
)
all_idx = set(players.index)
week_cols = [c for c in players.columns if c.startswith("week")]

players


# In[ ]:


# Define draft order
draft = [1, 2, 3, 3, 2, 1]
num_teams = len(set(draft))


# In[ ]:


# Define constraints
num_players = 2
pos_const = {"QB": 1, "RB": 1}


# In[ ]:


def display_draft(picks_idx):
    for team, idx in picks_idx.items():
        print(f"--- Team {team} ---")
        roster = players.loc[idx, ["uid"] + week_cols].set_index("uid")
        sum_points = roster.sum(axis=0).to_frame().rename({0: "sum_points"}, axis=1).T
        sum_points["min"] = sum_points.min(axis=1)
        display(roster)
        display(sum_points.sort_index(axis=1))


# # Method 1: Naive
# Optimize as if you get all picks in a row.

# In[ ]:


# Keep track of picks
picks_idx = {
    1: [],
    2: [],
    3: [],
}
picked_idx = []

# Optimize draft
for team in draft:
    # Get available players and those already picked by the team
    available_idx = all_idx - set(picked_idx)
    prev_picks_idx = set(picks_idx[team])
    available_idx |= prev_picks_idx
    available_idx = sorted(available_idx)

    # Get data
    uid_vals = players.loc[available_idx, "uid"].values
    pos_vals = players.loc[available_idx, "position"].values
    points_vals = players.loc[available_idx, week_cols].values

    # The variable we are solving for. We define our output variable as a bool
    # since we have to make a binary decision on each player (pick or don't pick)
    roster = cp.Variable(len(available_idx), boolean=True)

    # Save constraints
    constraints = []

    # Our roster must be composed of exactly `num_players` players
    constraints.append(cp.sum(roster) == num_players)

    # Define position constraints
    for pos, num in pos_const.items():
        is_pos = pos_vals == pos
        constraints.append(cp.sum(is_pos @ roster) == num)

    # Define constraints corresponding already picked players
    for idx in prev_picks_idx:
        did_pick = uid_vals == players.loc[idx, "uid"]
        constraints.append(cp.sum(did_pick @ roster) == 1)

    # Define the objective
    weekly_points = roster @ points_vals
    min_weekly_points = cp.min(weekly_points)
    objective = cp.Maximize(min_weekly_points)

    # Solve
    problem = cp.Problem(objective, constraints)
    problem.solve(max_iters=25)

    # Get result
    roster_idx = roster.value
    opt_roster = np.array(available_idx)[roster_idx.astype(bool)]  # re-align indices
    new_pick_idx = list(set(opt_roster) - prev_picks_idx)[0]  # TODO: use ADP or something instead of first
    picks_idx[team].append(new_pick_idx)
    picked_idx.append(new_pick_idx)

# Display draft
display_draft(picks_idx)
```

**Reply by SteveDiamond** [reply]
I posted your problem in the discord: https://discord.gg/4urRQeGBCr
That's the best place to go to discuss with the community.

**Comment by kmcentush** [comment]
One suggestion in the Discord was to look at [OWAs](https://en.wikipedia.org/wiki/Ordered_weighted_averaging_aggregation_operator) in the hope that my objective can be re-cast into that scheme. I think it's possible in principle, but I don't have a strong enough background in math to be confident. Below is an attempt to set up the OWA problem - I recognize that I'm definitely not using the weights concept correctly, nor at any point am I actually apply any OWA operation. Any help would be appreciated!

Note: this code is an extension from the previous post.
```
# Compute weights
# TODO: what should these be? is the dimensionality even correct?
n = num_teams
step_size = 1 / ((n * (n + 1)) / 2)
weights = np.arange(n, 0, -1).reshape(-1, 1) * step_size
# weights = (np.ones(num_teams) / np.arange(1, n + 1)).reshape(-1, 1)
weights

# Get data
uid_vals = players["uid"].values
pos_vals = players["position"].values
points_vals = players[week_cols].values

# The variable we are solving for. We define our output variable as a bool
# since we have to make a binary decision on each player (pick or don't pick)
roster = cp.Variable(shape=(num_teams, len(set(uid_vals))), boolean=True)

# Save constraints
constraints = []

# Each roster must be composed of exactly `num_players` players
constraints.append(cp.sum(roster, axis=1) == num_players)

# A player can only be picked at most once
constraints.append(cp.sum(roster, axis=0) <= 1)

# Define position constraints per team
for pos, num in pos_const.items():
    is_pos = pos_vals == pos
    constraints.append(roster @ is_pos == num)
    
# Define the objective
weekly_points = roster @ points_vals
min_weekly_points = cp.min(weekly_points, axis=1)  # per team
objective = cp.Maximize(weights.T @ min_weekly_points)  # this isn't correct; how do I incorporate weights?

# Solve
problem = cp.Problem(objective, constraints)
problem.solve(max_iters=25)

# Get result
roster_idx = roster.value
picks_idx = {}
for i in range(roster_idx.shape[0]):
    picks_idx[i + 1] = np.argwhere(roster_idx[i] == 1).ravel()

# Display draft
display_draft(picks_idx)
```

Now, this code does actually yield a solution but it does not match my dummy example - indicating that the objective is not correctly defined with my goal in mind.

**Comment by kmcentush** [comment]
I managed to achieve what I was after by combining my method 1 with decision trees - in essence having one convex problem/solution for each node. An interesting aside - it found a more optimal draft pattern for team 1 than I gave in my example above. I've attached my notebook for anyone to take a look.

**Main question:** Are there any tips to get a solver to run faster (other than improved CPU)? I don't know if there's something like JAX or GPU support, etc. that can help.

In even my simple dummy case, I can have 500+ nodes that each take ~0.01 seconds to solve, or ~5s for the entire tree. Sure that seems fast, but the scale will obviously blow up when I expand beyond my limited dummy example. My dummy draft has 3 teams and 2 picks each. A real draft has 12 teams and and 17 picks each. Plus, there about two orders of magnitudes more players for the optimizer to consider (more positions, more choices per position) and constrain, so I imagine the ~0.01 seconds I cited on my machine will go out the window in this case as well. All-in-all, I imagine my approach is quite limited in practice due to the nature of the wide tree and convex optimization that happens at each node.

**Secondary question:** Does anyone have any other tips? I would still love to figure out if there's a way to do this all with a single convex optimization.

Code:
[dummy_optimizer_v2.zip](https://github.com/cvxpy/cvxpy/files/9215942/dummy_optimizer_v2.zip)

**Reply by rileyjmurray** [reply]
Thanks for the update! This is a really cool example.

As to getting things to run faster, you could use Dask. Specifically, ``dask.delayed`` lets you build up a task graph that's only resolved when some output is needed, and it can manage parallelism for any problems that can be solved in an embarrassingly parallel way. This is mostly useful if you're solving small problems with a single-threaded solver like ECOS (which is indeed extremely fast on small problems).

Another possibility for getting things to run faster would depend on (1) if your problems only differ in values specified by CVXPY Parameter objects and (2) if those problems were DPP. If that's the case then you can use SCS with warm-starting. The caveat here is that the convergence of SCS can be much slower than ECOS, so you'd really be banking on the warm-starting making a big difference.

CVXPY doesn't have any meaningful GPU support at the moment. SCS can technically be installed so it uses a GPU for part of its algorithm, but I haven't seen big improvements from that when I tried it several years ago. (Maybe my GPU was too old.)

**Reply by kmcentush** [reply]
Thanks @rileyjmurray!

I gave Dask a shot but ran into some issues given how I constructed the node tree - basically, the parallel processes were implicitly copying nodes rather than passing around the same instance (no surprise - parallelization isn't able to pass exact instances around by design), so I was getting really weird behaviors when traversing the tree's parent <-> children relationships.

I imagine there is a way for me to take advantage of Dask if I redesign the nodes/tree with this behavior in mind (ex: hashmap/dict of all nodes keyed by a UID (some hash of the node's content), and then making the parent <-> children relationships go through this rather than the objects themselves), but I'll leave that for another day.

Do you have any suggestions with respect to casting my problem into a single optimization? Is there some way to exploit constraints to dictate "priority"?

**Reply by rileyjmurray** [reply]
Happy to help! Sadly, I don't have any more specific suggestions about how to cast this as a single optimization problem. The modeling that you're doing is really impressive. My one suggestion would be to consider how mixed-integer modeling might be of use to you. If you keep the number of integer variables small then you should still be able to solve the problem quickly. We have an open pull request about adding support for a nice mixed-integer linear programming solver available in the latest SciPy. So if you come up with an integer programming formulation but need a good open-source mixed-integer solver then you should be able to use the latest version of cvxpy on the master branch.
